lib/paper_trail.ex

```diff
@@ -85,65 +76,55 @@ defmodule PaperTrail do
-    repo = PaperTrail.RepoClient.repo()
-    client = PaperTrail.RepoClient
-
-    repo.transaction(fn ->
-      case client.strict_mode() do
-        true ->
-          version_data =
-            changeset.data
-            |> Map.merge(%{
-              current_version_id: get_sequence_id("versions")
-            })
-
-          target_changeset = changeset |> Map.merge(%{data: version_data})
-          target_version = make_version_struct(%{event: "update"}, target_changeset, options)
-          initial_version = repo.insert!(target_version)
-          updated_changeset = changeset |> change(%{current_version_id: initial_version.id})
-          model = repo.update!(updated_changeset)
-
-          new_item_changes =
-            initial_version.item_changes
-            |> Map.merge(%{
-              current_version_id: initial_version.id
-            })
-
-          initial_version |> change(%{item_changes: new_item_changes}) |> repo.update!
-          model
-
-        _ ->
-          model = repo.update!(changeset)
-          version_struct = make_version_struct(%{event: "update"}, changeset, options)
-          repo.insert!(version_struct)
-          model
-      end
-    end)
-    |> elem(1)
+    changeset
+    |> update(options)
+    |> model_or_error()
   end
 
   @doc """
   Deletes a record from the database with a related version insertion in one transaction
   """
-  def delete(struct, options \\ [origin: nil, meta: nil, originator: nil, prefix: nil]) do
+  @spec delete(model_or_changeset :: model | Ecto.Changeset.t(model), options :: Keyword.t()) ::
+          {:ok, %{model: model, version: Version.t()}} | {:error, Ecto.Changeset.t(model) | term}
+        when model: struct
+  def delete(
+        model_or_changeset,
+        options \\ [origin: nil, meta: nil, originator: nil, prefix: nil]
+      ) do
     PaperTrail.Multi.new()
-    |> PaperTrail.Multi.delete(struct, options)
+    |> PaperTrail.Multi.delete(model_or_changeset, options)
     |> PaperTrail.Multi.commit()
   end
 
   @doc """
   Same as delete/2 but returns only the model struct or raises if the changeset is invalid.
   """
-  def delete!(struct, options \\ [origin: nil, meta: nil, originator: nil, prefix: nil]) do
-    repo = PaperTrail.RepoClient.repo()
+  @spec delete!(model_or_changeset :: model | Ecto.Changeset.t(model), options :: Keyword.t()) ::
+          model
+        when model: struct
+  def delete!(
+        model_or_changeset,
+        options \\ [origin: nil, meta: nil, originator: nil, prefix: nil]
+      ) do
+    model_or_changeset
+    |> delete(options)
+    |> model_or_error()
+  end
+
+  @spec model_or_error({:ok, %{model: model}}) :: model when model: struct()
+  defp model_or_error({:ok, %{model: model}}) do
+    model
+  end
+
+  @spec model_or_error({:error, reason :: term}) :: no_return
+  defp model_or_error({:error, %Ecto.Changeset{} = changeset}) do
+    raise Ecto.InvalidChangesetError, action: :update, changeset: changeset
```

Comment:
previously I was using default Ecto Repo `insert!`, `update!`, `delete!` errors this one seemed strange to me the action is always action: :update:

`action:  :update`. Is this the default behavior of Ecto? I'll also look this up.

Reply:
Hm, no. I think I did it wrong:

https://github.com/elixir-ecto/ecto/blob/v3.5.5/lib/ecto/repo/schema.ex#L195

I'll push a fix tomorrow.

Reply:
Handeled in this PR: https://github.com/izelnakri/paper_trail/pull/124#discussion_r543383763